### PR TITLE
Add active account option in msal-browser

### PR DESCRIPTION
**Pull request by pkanher617**
This PR adds the ability to set an active account for an instance of MSAL. This account MUST be set by the application using the MSAL library, and will be used when no account is given to any of our APIs. 

This account is cleared when logout is called, either with no account or a matching account. If a non-matching account is passed to logout, the active account is kept intact.

### CODE REVIEW

File: lib/msal-browser/src/app/ClientApplication.ts [hunk 1]
```diff
@@ -49,6 +49,9 @@ export abstract class ClientApplication {
     // Flag to indicate if in browser environment
     protected isBrowserEnvironment: boolean;
 
+    // Sets the account to use if no account info is given
+    private activeAccountId: string | null;
```
Comment (jasonnutter) [inline lib/msal-browser/src/app/ClientApplication.ts:53]: I think we shuold consider making this name reflect which id is used:

```suggestion
    private activeLocalAccountId: string | null;
```

File: lib/msal-browser/src/app/IPublicClientApplication.ts [hunk 2]
```diff
@@ -28,6 +28,8 @@ export interface IPublicClientApplication {
     ssoSilent(request: SsoSilentRequest): Promise<AuthenticationResult>;
     getLogger(): Logger;
     setLogger(logger: Logger): void;
+    setActiveAccount(account: AccountInfo): void;
+    getActiveAccount(): AccountInfo;
```
Comment (tnorling) [inline lib/msal-browser/src/app/IPublicClientApplication.ts:32]:
```suggestion
    getActiveAccount(): AccountInfo | null;
```

File: lib/msal-browser/src/app/ClientApplication.ts [hunk 3]
```diff
@@ -644,6 +652,29 @@ export abstract class ClientApplication {
         }
     }
 
+    /**
+     * Sets the account to use as the active account. If no account is passed to the acquireToken APIs, then MSAL will use this active account.
+     * @param account 
+     */
+    setActiveAccount(account: AccountInfo): void {
```
Comment (tnorling) [inline lib/msal-browser/src/app/ClientApplication.ts:659]:
```suggestion
    setActiveAccount(account: AccountInfo | null): void {
```

File: lib/msal-browser/src/app/IPublicClientApplication.ts [hunk 4]
```diff
@@ -28,6 +28,8 @@ export interface IPublicClientApplication {
     ssoSilent(request: SsoSilentRequest): Promise<AuthenticationResult>;
     getLogger(): Logger;
     setLogger(logger: Logger): void;
+    setActiveAccount(account: AccountInfo): void;
```
Comment (tnorling) [inline lib/msal-browser/src/app/IPublicClientApplication.ts:31]:
```suggestion
    setActiveAccount(account: AccountInfo | null): void;
```

File: lib/msal-browser/src/app/ClientApplication.ts [hunk 5]
```diff
@@ -644,6 +652,29 @@ export abstract class ClientApplication {
...
+        if (!account) {
+            this.activeAccountId = null;
+            return;
+        }
+        this.activeAccountId = account.localAccountId;
```
Comment (jasonnutter) [inline lib/msal-browser/src/app/ClientApplication.ts:664]: Nitpick:

```suggestion
        this.activeAccountId = account ? account.localAccountId : null;
```

File: lib/msal-browser/src/app/ClientApplication.ts (hunk 1 above)
Comment (pkanher617) [reply lib/msal-browser/src/app/ClientApplication.ts]: done.

File: lib/msal-browser/src/app/IPublicClientApplication.ts (hunk 2 above)
Comment (pkanher617) [reply lib/msal-browser/src/app/IPublicClientApplication.ts]: done.

File: lib/msal-browser/src/app/ClientApplication.ts (hunk 3 above)
Comment (pkanher617) [reply lib/msal-browser/src/app/ClientApplication.ts]: done.

File: lib/msal-browser/src/app/IPublicClientApplication.ts (hunk 4 above)
Comment (pkanher617) [reply lib/msal-browser/src/app/IPublicClientApplication.ts]: done.

File: lib/msal-browser/src/app/ClientApplication.ts (hunk 5 above)
Comment (pkanher617) [reply lib/msal-browser/src/app/ClientApplication.ts]: done.

File: lib/msal-common/src/cache/entities/AccountEntity.ts [hunk 6]
```diff
@@ -275,4 +275,21 @@ export class AccountEntity {
...
     }
+
+    /**
+     * Helper function to determine whether 2 arrays are equal
+     * Used to avoid unnecessary state updates
+     * @param arrayA 
+     * @param arrayB 
+     */
+    static accountInfoIsEqual(accountA: AccountInfo, accountB: AccountInfo): boolean {
+        if (!accountA || !accountB) {
```
Comment (tnorling) [inline lib/msal-common/src/cache/entities/AccountEntity.ts:286]: What if they're both null?

**Review by tnorling — APPROVED** [review]

File: lib/msal-angular/package-lock.json [hunk 7]
```diff
@@ -420,24 +420,6 @@
...
-    },
-    "@azure/msal-common": {
-      "version": "2.0.0",
-      "resolved": "https://registry.npmjs.org/@azure/msal-common/-/msal-common-2.0.0.tgz",
-      "integrity": "sha512-d1RNcJb+P1EGzMHtgbZoVlHLQWjlVfr504jywNk9YEfoq8Hw3BxJ0wepu+1w0hc64D8zG0wljcvHaIH1jTn2SA==",
-      "dev": true,
-      "requires": {
-        "debug": "^4.1.1"
-      }
-    },
```
Comment (jasonnutter) [inline lib/msal-angular/package-lock.json:440]: Can you revert this?

**Review by jasonnutter — APPROVED** [review]

**Comment by jasonnutter** [comment]
FYI @jo-arroyo once this gets merged, we should update the MSAL Angular v2 interceptor to use the active account (potentially falling back to the first account in getAllAccounts).

File: lib/msal-common/src/cache/entities/AccountEntity.ts [hunk 8]
```diff
@@ -275,4 +275,21 @@ export class AccountEntity {
...
+     */
+    static accountInfoIsEqual(accountA: AccountInfo | null, accountB: AccountInfo | null): boolean {
+        if (!accountA || !accountB) {
+            return false;
+        }
+        return (accountA.homeAccountId === accountB.homeAccountId) && 
+            (accountA.localAccountId === accountB.localAccountId) &&
+            (accountA.username === accountB.username) &&
+            (accountA.tenantId === accountB.tenantId) &&
+            (accountA.environment === accountB.environment);
```
Comment (sameerag) [inline lib/msal-common/src/cache/entities/AccountEntity.ts:293]: Shouldn't we respect environment aliases here?

Comment (pkanher617) [reply lib/msal-common/src/cache/entities/AccountEntity.ts:293]: For now this is just an exact match. It is only being used to clear the active account. If we want to add environment aliasing for other scenarios we can, but I think for logout it is fine.

Comment (sameerag) [reply lib/msal-common/src/cache/entities/AccountEntity.ts:293]: okay!

**Review by sameerag — APPROVED** [review]
Lgtm. Please note the aliases change in case this use case is extended later.

**Pull request merged**